File: leetcode/Python/code/220.存在重复元素/220.存在重复元素.py
from typing import Generator, List


class Solution:
    def containsNearbyAlmostDuplicate(
        self, nums: List[int], indexDiff: int, valueDiff: int
    ) -> bool:
        if len(nums) <= 1 or indexDiff < 0 or valueDiff < 0:
            return False

        # Hash, 处理每个桶
        all_buckets = {}

        # 每个桶的容量
        bucket_size = valueDiff + 1

        for i, v in enumerate(nums):
            # 1. 取商, 决定放入哪个桶
            bucket_num = v // bucket_size

            # 2. 桶中已经有元素了: 说明记录相同桶的上一个值 和 本次的值 都在valueDiff 范围内。
            # 同时, if i >= indexDiff: 语句保证了相同桶的索引差值始终在 indexDiff 范围内
            if bucket_num in all_buckets:
                return True

            # 3. 比较上一个桶
            if (bucket_num - 1) in all_buckets and abs(
                all_buckets[bucket_num - 1] - v
            ) <= valueDiff:
                return True

            # 4. 比较下一个桶
            if (bucket_num + 1) in all_buckets and abs(
                all_buckets[bucket_num + 1] - v
            ) <= valueDiff:
                return True

            # 5. 将值放进第bucket_num桶内
            all_buckets[bucket_num] = v  # 把v放入桶中

            # 6. 滑动窗口.
            # 如果不构成返回条件，那么当i >= indexDiff 的时候就要删除旧桶了，
            # 以维持桶中的元素索引跟下一个i+1索引只差不超过indexDiff
            # !!!: 保证了 all_buckets 的数量始终维持在 indexDiff 内。
            if i >= indexDiff:
                all_buckets.pop(nums[i - indexDiff] // bucket_size)

        return False


s = Solution()
res = s.containsNearbyAlmostDuplicate([1, 2, 3, 1], 3, 0)
print(res)

File: leetcode/Go/code/220.存在重复元素/220.存在重复元素.go
package main

func divRoundDown(numerator, denominator int) int {
	quotient := numerator / denominator
	if numerator%denominator != 0 && (numerator < 0) != (denominator < 0) {
		quotient--
	}
	return quotient
}

func abs(x int) int {
	if x < 0 {
		return -x
	}

	return x
}

func containsNearbyAlmostDuplicate(nums []int, indexDiff int, valueDiff int) bool {
	if len(nums) <= 1 || indexDiff < 0 || valueDiff < 0 {
		return false
	}

	allBuckets := make(map[int]int, indexDiff+1)

	// Set the size of bucket.
	bucketSize := valueDiff + 1

	for i, v := range nums {
		// 1. Get the id of bucket
		bucketId := divRoundDown(v, bucketSize)

		// 2. If the bucketId is already existed, return true.
		if _, ok := allBuckets[bucketId]; ok {
			return true
		}

		// 3. Compare to previous bucket
		if _, ok := allBuckets[bucketId-1]; ok && abs(allBuckets[bucketId-1]-v) <= valueDiff {
			return true
		}

		// 4. Compare to next bucket
		if _, ok := allBuckets[bucketId+1]; ok && abs(allBuckets[bucketId+1]-v) <= valueDiff {
			return true
		}

		// 5. Put the bucketId into allBuckets with value v.
		allBuckets[bucketId] = v

		// 6. If the current index is great than indexDiff,
		// then pop the i - indexDiff element in nums which match bucketId in allBuckets.
		if i >= indexDiff {
			delete(allBuckets, divRoundDown(nums[i-indexDiff], bucketSize))
		}
	}

	return false
}

func main() {
	res := containsNearbyAlmostDuplicate([]int{1, 5, 9, 1, 5, 9}, 2, 3)
	println(res)
}
